pushsync/simulation_test.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pushsync
+
+import (
+	"context"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/retrieval"
+	"github.com/ethersphere/swarm/network/simulation"
+	"github.com/ethersphere/swarm/pss"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	bucketKeyPushSyncer = simulation.BucketKey("pushsyncer")
+	bucketKeyNetStore   = simulation.BucketKey("netstore")
+)
+
+var (
+	nodeCntFlag   = flag.Int("nodes", 4, "number of nodes in simulation")
+	chunkCntFlag  = flag.Int("chunks", 4, "number of chunks per upload in simulation")
+	testCasesFlag = flag.Int("cases", 4, "number of concurrent upload-download cases to test in simulation")
+)
+
+// test syncer using pss
+// the test
+// * creates a simulation with connectivity loaded from a snapshot
+// * for each test case, two nodes are chosen randomly, an uploader and a downloader
+// * uploader uploads a number of chunks
+// * wait until the uploaded chunks are synced
+// * downloader downloads the chunk
+// Testcases are run concurrently
+func TestPushsyncSimulation(t *testing.T) {
```

Comment:
This test is failing for two reasons:
- snapshot loading failing as no peers and connections are in kademlia (WaitTillSnapshotRecreated simulations method is requiring kademlia connections) which means that we must always have bzz protocol for simulations services
- netstore missing RemoteGet function

These changes are solving these problems:

```diff
diff --git a/pushsync/simulation_test.go b/pushsync/simulation_test.go
index bf50748cd..5b2bffe75 100644
--- a/pushsync/simulation_test.go
+++ b/pushsync/simulation_test.go
@@ -200,6 +200,7 @@ func newServiceFunc(ctx *adapters.ServiceContext, bucket *sync.Map) (node.Servic
        bucket.Store(bucketKeyNetStore, netStore)
 
        r := retrieval.New(kad, netStore, kad.BaseAddr())
+       netStore.RemoteGet = r.RequestFromPeers
 
        pubSub := pss.NewPubSub(ps)
        // setup pusher
@@ -216,13 +217,20 @@ func newServiceFunc(ctx *adapters.ServiceContext, bucket *sync.Map) (node.Servic
                os.RemoveAll(dir)
        }
 
-       return &RetrievalAndPss{r, ps}, cleanup, nil
+       bzz := network.NewBzz(&network.BzzConfig{
+               OverlayAddr:  addr.Over(),
+               UnderlayAddr: addr.Under(),
+               HiveParams:   network.NewHiveParams(),
+       }, kad, nil, nil, nil, nil, nil)
+
+       return &RetrievalAndPss{r, ps, bzz}, cleanup, nil
 }
 
 // implements the node.Service interface
 type RetrievalAndPss struct {
        retrieval *retrieval.Retrieval
        pss       *pss.Pss
+       bzz       *network.Bzz
 }
 
 func (s *RetrievalAndPss) APIs() []rpc.API {
@@ -230,11 +238,15 @@ func (s *RetrievalAndPss) APIs() []rpc.API {
 }
 
 func (s *RetrievalAndPss) Protocols() []p2p.Protocol {
-       return append(s.retrieval.Protocols(), s.pss.Protocols()...)
+       return append(append(s.retrieval.Protocols(), s.pss.Protocols()...), s.bzz.Protocols()...)
 }
 
 func (s *RetrievalAndPss) Start(srv *p2p.Server) error {
-       err := s.retrieval.Start(srv)
+       err := s.bzz.Start(srv)
+       if err != nil {
+               return err
+       }
+       err = s.retrieval.Start(srv)
        if err != nil {
                return err
        }
@@ -242,7 +254,11 @@ func (s *RetrievalAndPss) Start(srv *p2p.Server) error {
 }
 
 func (s *RetrievalAndPss) Stop() error {
-       err := s.retrieval.Stop()
+       err := s.bzz.Stop()
+       if err != nil {
+               return err
+       }
+       err = s.retrieval.Stop()
        if err != nil {
                return err
        }
```